handlers/public.go

```diff
@@ -0,0 +1,208 @@
+	}
+
+	requestDecoder := json.NewDecoder(req.Body)
+	var b createPaymentResource
+	err := requestDecoder.Decode(&b)
+	if err != nil {
+		log.ErrorR(req, fmt.Errorf("request body invalid"))
+		w.WriteHeader(http.StatusBadRequest) // 400
+		return
+	}
+
+	resource := b.Resource
+
+	cfg, err := config.Get()
+	if err != nil {
+		log.ErrorR(req, fmt.Errorf("error getting config: [%s]", err))
+		w.WriteHeader(http.StatusInternalServerError) // 500
+		return
+	}
+	url, err := url.Parse(resource)
+	if err != nil {
+		log.ErrorR(req, fmt.Errorf("error parsing resource: [%s]", err))
+		w.WriteHeader(http.StatusBadRequest) // 400
+		return
+	}
+
+	whitelist := strings.Split(cfg.DomainWhitelist, ",")
+	matched := false
+	for _, domain := range whitelist {
+		if url.Host == domain {
+			matched = true
+			break
+		}
+	}
+	if !matched {
+		log.ErrorR(req, fmt.Errorf("invalid resource domain: %s", url.Host))
+		w.WriteHeader(http.StatusBadRequest) // 400
+		return
+	}
+
+	resourceReq, err := http.NewRequest("GET", resource, nil)
+	if err != nil {
+		log.ErrorR(resourceReq, fmt.Errorf("failed to create Resource Request: [%s]", err))
+		w.WriteHeader(http.StatusInternalServerError) // 500
+		return
+	}
+	resp, err := client.Do(resourceReq)
+	if err != nil {
+		log.ErrorR(resourceReq, fmt.Errorf("error getting Cost Resource: [%s]", err))
+		w.WriteHeader(http.StatusInternalServerError) // 500
+		return
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.ErrorR(resourceReq, fmt.Errorf("error reading Cost Resource: [%s]", err))
+		w.WriteHeader(http.StatusInternalServerError) // 500
+		return
+	}
+
+	// TODO save cost resource and ensure all mandatory fields are present:
+
+	paymentResource := &PaymentResource{}
+	err = json.Unmarshal(body, paymentResource)
+	if err != nil {
+		log.ErrorR(resourceReq, fmt.Errorf("error reading Cost Resource: [%s]", err))
+		w.WriteHeader(http.StatusInternalServerError) // 500
+		return
+	}
+
+	user := strings.Split(req.Header.Get("Eric-Authorised-User"), ";")
+	email := user[0]
+	var forename string
+	var surname string
+
+	for i := 1; i < len(user); i++ {
+		v := strings.Split(user[i], "=")
+		if v[0] == " forename" {
+			forename = v[1]
+		} else if v[0] == " surname" {
+			surname = v[1]
+		} else {
+			log.ErrorR(req, fmt.Errorf("unexpected format in Eric-Authorised-User: %s", user))
+			w.WriteHeader(http.StatusInternalServerError) // 500
+			return
+		}
+	}
+	paymentResource.CreatedBy = CreatedBy{
+		ID:       req.Header.Get("Eric-Identity"),
+		Email:    email,
+		Forename: forename,
+		Surname:  surname,
+	}
+
+	paymentResource.CreatedAt = time.Now()
+	paymentResource.Reference = b.Reference
+
+	// Write to DB
+	session, err := data.GetMongoSession()
+	if err != nil {
+		log.ErrorR(req, fmt.Errorf("error connecting to MongoDB: %s", err))
+		w.WriteHeader(http.StatusInternalServerError) // 500
+		return
+	}
+	defer session.Close()
+
+	c := session.DB("transactions").C("payments")
```

Comment:
just wondering, im assuming `C` stands for collections? Could change it so it has `collections`? Or is this more work than it seems?

Reply:
@ChrisJBurns The GoDoc is [here](https://godoc.org/gopkg.in/mgo.v2#Database.C). The variable name could be changed to "collections" instead of "c" but the function name is in the mgo package.

Reply:
Ahh i see, its a Mongo library. ignore me.